src/data/map.rs

```diff
@@ -0,0 +1,1561 @@
+    /// Deletes an entry
+    Delete(SentriedKey),
+}
+
+/// Indicates whether the transaction can perform permanent deletion or not.
+#[derive(Hash, Eq, PartialEq, PartialOrd, Ord, Clone, Serialize, Deserialize)]
+pub enum MapTransaction {
+    /// Only soft deletes possible.
+    Commit(SentryOption),
+    /// Allows hard-deletes.
+    HardCommit(SentryOption),
+}
+
+/// Indicates whether the transaction is carried out with concurrency control or not.
+#[derive(Hash, Eq, PartialEq, PartialOrd, Ord, Clone, Serialize, Deserialize, Debug)]
+pub enum SentryOption {
+    /// No concurrency control.
+    AnyVersion(Transaction),
+    /// Optimistic concurrency.
+    ExpectVersion(SentriedTransaction),
+}
+
+// pub type Transaction = Vec<Cmd>;
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize, Default, Debug)]
+pub struct Transaction(Vec<Cmd>);
+
+impl Transaction {
+    pub fn get(&self) -> &Vec<Cmd> {
+        &self.0
+    }
+}
+
+impl Deref for Transaction {
+    type Target = Vec<Cmd>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl From<Vec<Cmd>> for Transaction {
+    fn from(vec: Vec<Cmd>) -> Self {
+        Transaction(vec)
+    }
+}
+
+// pub type SentriedTransaction = Vec<SentriedCmd>;
+///
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize, Default, Debug)]
+pub struct SentriedTransaction(Vec<SentriedCmd>);
+
+impl SentriedTransaction {
+    pub fn get(&self) -> &Vec<SentriedCmd> {
+        &self.0
+    }
+}
+
+impl Deref for SentriedTransaction {
+    type Target = Vec<SentriedCmd>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl From<Vec<SentriedCmd>> for SentriedTransaction {
+    fn from(vec: Vec<SentriedCmd>) -> Self {
+        SentriedTransaction(vec)
+    }
+}
+
+///
+pub type ExpectedVersion = u64;
+///
+pub type SentriedKey = (Key, ExpectedVersion);
+///
+pub type SentriedKvPair = (KvPair, ExpectedVersion);
+
+/// Common methods for NonSentried flavours.
+impl<P: AccessListTrait> MapBase<P, NonSentried> {
+    /// Commit transaction.
+    ///
+    /// If the specified `expected_version` does not equal the entries count in data, an
+    /// error will be returned.
+    pub fn commit(&mut self, tx: &Transaction) -> Result<()> {
+        // Deconstruct tx into inserts, updates, and deletes
+        let operations: Operations = tx.iter().fold(
+            Default::default(),
+            |(mut insert, mut update, mut delete), cmd| {
+                match cmd {
+                    Cmd::Insert(kv_pair) => {
+                        let _ = insert.insert(kv_pair.clone());
+                    }
+                    Cmd::Update(kv_pair) => {
+                        let _ = update.insert(kv_pair.clone());
+                    }
+                    Cmd::Delete(key) => {
+                        let _ = delete.insert(key.clone());
+                    }
+                };
+                (insert, update, delete)
```

Comment:
will this lose the sequence of interleaving?
for example `insert A, delete A, insert A, update A` will be applied as: insert A, then update A, then delete A ?

Reply:
Yes well spotted, this is a problem discussed in the pre-rfc here: https://github.com/maidsafe/pre-rfc/pull/112#discussion_r351410563
I suggest there that it should be fixed.

Reply:
Yes well spotted, this is a problem discussed in the pre-rfc.
I suggest there that it should be fixed:

> We would need to consider what behaviour we want to support, and adapt network side handling to it. If we want to be able to do this:
> 
> ```
> let mymap = new Map();
> mymap.startTx();
> mymap.insert("key1", "value1");
> mymap.delete("key1");
> mymap.insert("key1", "value2"); // <---- insert same key again
> mymap.commit();
> ```
> 
> or even
> 
> ```
> let mymap = new Map();
> mymap.startTx();
> mymap.insert("key1", "value1");
> mymap.delete("key1");
> mymap.insert("key1", "value1"); // <--  insert same key again, with same value.
> mymap.commit();
> ```
> 
> .. then we have to change some of the current logics.
> Since it applies all inserts in a group, then all updates and then all deletes, the first example would blow up since it tries to perform
> ```
> mymap.insert("key1", "value1");
> mymap.insert("key1", "value2");
> mymap.delete("key1");
> ```
> 
> But, that's not a big thing to fix, we would then group operations by key, and _then_ apply them in the order they were added to the tx.
> 
> The second example would not result in "key1" = "value1", even if we fixed the key-grouping as per above, since the last operation - with current logics - would be filtered out, so we would end up with the key being deleted. Also not a big thing to fix, but we just need to be aware of things like this :)
>